build-system/compile/compile.js

```diff
@@ -58,6 +62,86 @@ function convertPathsToTmpRoot(paths) {
   return paths.map(path => path.replace(/^(!?)(.*)$/, `$1${SRC_TEMP_DIR}/$2`));
 }
 
+function loadSourceMap(file) {
+  if (file.startsWith('dist')) {
+    return fs.readFile(`${file}.map`);
+  }
+  return null;
+}
+
+/**
+ * Apply Babel Transforms on output from Closure Compuler, then cleanup added space with Terser.
+ *
+ * @param {string} directory directory this file lives in
+ * @param {boolean} isEsmBuild
+ * @return {!Promise}
+ */
+function postClosureBabel(directory, isEsmBuild) {
+  return through.obj(function(file, enc, next) {
+    if (pathModule.extname(file.path) === '.map') {
+      return next(null, file);
+    }
+
+    const babelPlugins = conf.plugins({isPostCompile: true, isEsmBuild});
+    if (babelPlugins.length === 0) {
+      return next(null, file);
+    }
+
+    function returnMapFirst(map) {
+      let first = true;
+      return function(file) {
+        if (first) {
+          first = false;
+          return map;
+        }
+        return loadSourceMap(file);
+      };
+    }
+
+    const map = loadSourceMap(file.path);
+    const {code, map: babelMap} = babel.transformSync(file.contents, {
+      plugins: babelPlugins,
+      retainLines: true,
+      sourceMaps: true,
+      inputSourceMap: false,
+    });
+    let remapped = resorcery(
+      babelMap,
+      returnMapFirst(map),
+      !argv.full_sourcemaps
+    );
+
+    const {code: compressed, map: terserMap} = terser.minify(code, {
+      mangle: false,
+      compress: {
+        defaults: false,
+        unused: true,
+      },
+      output: {
+        beautify: !!argv.pretty_print,
+        comments: /\/*/,
+        keep_quoted_props: true,
+      },
+      sourceMap: true,
+    });
+    file.contents = Buffer.from(compressed, 'utf-8');
+
+    // TODO: Resorcery should support a chain, instead of having to call
+    // multiple times.
+    remapped = resorcery(
+      terserMap,
+      returnMapFirst(remapped),
+      !argv.full_sourcemaps
+    );
+    fs.writeFileSync(
+      pathModule.resolve(directory, `${pathModule.basename(file.path)}.map`),
+      remapped.toString()
+    );
+
+    return next(null, file);
+  });
+}
```

Comment:
This is a pretty substantial function. Could we move it into a separate file for easier maintenance in future?

Reply:
Done.